pkg/services/x509/x509_validator.go

```diff
@@ -0,0 +1,270 @@
+	leaf := j.chain[0]
+	var subjAltName pkix.Extension
+	for _, ext := range leaf.Extensions {
+		if ext.Id.Equal(subjectAltNameID) {
+			subjAltName = ext
+		}
+		break
+	}
+	if len(subjAltName.Value) == 0 {
+		return "", nil
+	}
+
+	var otherNames generalNames
+	if _, err := asn1.Unmarshal(subjAltName.Value, &otherNames); err != nil {
+		return "", err
+	}
+	var otherNameStr string
+	if _, err := asn1.Unmarshal(otherNames.OtherName.Value.Bytes, &otherNameStr); err != nil {
+		return "", err
+	}
+	return otherNameStr, nil
+}
+
+func NewJwtX509Validator(roots, intermediates []*x509.Certificate, allowedSigAlgs []jwa.SignatureAlgorithm) *JwtX509Validator {
+	return &JwtX509Validator{
+		roots:              roots,
+		intermediates:      intermediates,
+		allowedSigningAlgs: allowedSigAlgs,
+	}
+}
+
+// Parse attempts to parse a string as a jws. It checks if the x5c header contains at least 1 certificate.
+// The signature should be signed with the private key of the leaf certificate.
+// No other validations are performed. Call Verify to verify the auth token.
+func (validator JwtX509Validator) Parse(rawAuthToken string) (*JwtX509Token, error) {
+	// check the cert chain in the jwt
+	rawHeader, _, _, err := jws.SplitCompact(bytes.NewReader([]byte(rawAuthToken)))
+	if err != nil {
+		return nil, fmt.Errorf("the jwt should contain of 3 parts: %w", err)
+	}
+
+	headers := jws.NewHeaders()
+	headerStr, _ := base64.RawURLEncoding.DecodeString(string(rawHeader))
+	if err = json.Unmarshal(headerStr, headers); err != nil {
+		return nil, fmt.Errorf("could not unmarshall headers: %w", err)
+	}
+
+	certsFromHeader := headers.X509CertChain()
+	if len(certsFromHeader) == 0 {
+		return nil, fmt.Errorf("the jwt x5c field should contain at least 1 certificate")
+	}
+
+	chain, err := validator.parseCertsFromHeader(certsFromHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	certUsedForSigning := chain[0]
+
+	payload, err := jws.Verify([]byte(rawAuthToken), headers.Algorithm(), certUsedForSigning.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify jwt signature: %w", err)
+	}
+
+	parsedJwt := jwt.New()
+	if err := json.Unmarshal(payload, &parsedJwt); err != nil {
+		return nil, fmt.Errorf("could not parse jwt payload: %w", err)
+	}
+
+	return &JwtX509Token{
+		chain:  chain,
+		token:  parsedJwt,
+		raw:    rawAuthToken,
+		sigAlg: headers.Algorithm(),
+	}, nil
+}
+
+func (validator JwtX509Validator) parseCertsFromHeader(certsFromHeader []string) ([]*x509.Certificate, error) {
+	// Parse all the certificates from the header into a chain
+	chain := []*x509.Certificate{}
+
+	for _, certStr := range certsFromHeader {
+		rawCert, err := base64.StdEncoding.Strict().DecodeString(certStr)
+		if err != nil {
+			return nil, fmt.Errorf("could not base64 decode cert: %w", err)
+		}
+		cert, err := x509.ParseCertificate(rawCert)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse x509 certificate: %w", err)
```

Comment:
consistency in error messages; on row 132 you're saying 'cert', here it's 'certificate'. It pushes my buttons :-p